### PR TITLE
Revamp "AnimationContext" Docs, now called "AnimTaskQueue"

### PR DESCRIPTION
**Pull request by fig02**
I got sent down a rabbit hole while doing Player docs which required a better understanding of what we used to call "AnimationContext".

After learning about it, I decided the current names for basically all of this system could be improved. After a few iterations I ended up calling it the "AnimTaskQueue" and changed a bunch of terminology throughout.
It sounds quite generic, but it kinda has to be. The various tasks aren't very correlated, other than the fact that they are related to animations.

All of the relevant functions in skelanime have comments at the top that hopefully explain everything, but let me know if anything else needs to be clarified.

### CODE REVIEW

File: include/z64animation.h [hunk 1]
```diff
@@ -239,95 +239,95 @@ s16 Animation_GetLength(void* animation);
...
-    /* 0x001 */ u8 vecCount;
-    /* 0x004 */ Vec3s* dst;
-    /* 0x008 */ Vec3s* src;
-    /* 0x00C */ u8* copyFlag;
-} AnimEntryCopyTrue; // size = 0x10
+    /* 0x00 */ u8 group;
+    /* 0x01 */ u8 vecCount;
+    /* 0x04 */ Vec3s* dest;
+    /* 0x08 */ Vec3s* src;
+    /* 0x0C */ u8* copyMap;
```
Comment (cadmic) [inline include/z64animation.h:280]: `limbSet`? This looks to me like an array of booleans indexed by limb denoting which set of limbs to copy

File: include/z64animation.h [hunk 2]
```diff
@@ -239,95 +239,95 @@ s16 Animation_GetLength(void* animation);
...
+    /* 0x38 */ OSMesg msg;
+} AnimTaskLoadPlayerFrame; // size = 0x3C
 
 typedef struct {
-    /* 0x000 */ u8 queueFlag;
-    /* 0x001 */ u8 vecCount;
-    /* 0x004 */ Vec3s* dst;
-    /* 0x008 */ Vec3s* src;
-} AnimEntryCopyAll; // size = 0xC
+    /* 0x00 */ u8 group;
```
Comment (cadmic) [inline include/z64animation.h:261]: `group` makes it sound like this is an id, but I was surprised to find that this is a bitset. Maybe `groups` or `groupFlags`?

File: src/code/z_skelanime.c [hunk 3]
```diff
@@ -798,245 +795,299 @@ void SkelAnime_InterpFrameTable(s32 limbCount, Vec3s* dst, Vec3s* start, Vec3s*
     }
 }
 
+static u32 sDisabledTransformTaskGroups = 0;
+static u32 sCurAnimTaskGroup;
+
 /**
- * zeroes out the current request count
+ * Resets the AnimTaskQueue.
+ * Zeroing the task count will stop the queue from processing tasks until new ones are added.
```
Comment (cadmic) [inline src/code/z_skelanime.c:803]: Nit: I'd just remove this sentence, it made me doubt whether the old tasks were "really" gone. Maybe `Resets the AnimTaskQueue.` -> `Removes all tasks from the AnimTaskQueue.`

File: src/code/z_skelanime.c [hunk 4]
```diff
@@ -798,245 +795,299 @@ void SkelAnime_InterpFrameTable(s32 limbCount, Vec3s* dst, Vec3s* start, Vec3s*
...
 }
 
 /**
- * Shifts the queue flag to the next queue
+ * Changes `sCurAnimTaskGroup` to the next group number.
+ *
+ * Task groups allow for disabling "transformative" tasks for a defined group.
+ * For more information see `AnimTaskQueue_DisableTransformTasksForGroup`.
+ *
+ * Note that `sCurAnimTaskGroup` is not a whole number that increments, it is handled at the bit-level.
```
Comment (cadmic) [inline src/code/z_skelanime.c:815]: I'd prefer this comment to go on `sCurAnimTaskGroup` instead (something like that this is a bitset where exactly one bit it set for the current group)

File: include/z64animation.h [hunk 5]
```diff
@@ -239,95 +239,95 @@ s16 Animation_GetLength(void* animation);
...
-    /* 0x004 */ Vec3s* dst;
-    /* 0x008 */ Vec3s* src;
-    /* 0x00C */ u8* copyFlag;
-} AnimEntryCopyFalse; // size = 0x10
+    /* 0x00 */ u8 group;
+    /* 0x01 */ u8 vecCount;
+    /* 0x04 */ Vec3s* dest;
+    /* 0x08 */ Vec3s* src;
+    /* 0x0C */ u8* copyMap;
+} AnimTaskCopyUsingMapInverted; // size = 0x10
```
Comment (cadmic) [inline include/z64animation.h:289]: `Inverted` makes me think that this inverts the limb orientations or something. Maybe something like set complement? That could also be confusing though.

File: include/z64animation.h (hunk 2 above)
Comment (fig02) [reply include/z64animation.h:261]: not sure if this became clear as you read more of the code, but its always going to be just one group number. it is handled with bits yes, but this value within each task will never hold multiple groups/bit numbers. it is a single group number that increments as "1, 2, 4, 8..." instead of "1, 2, 3..."

File: src/code/z_skelanime.c (hunk 3 above)
Comment (fig02) [reply src/code/z_skelanime.c:803]: The same reaction you have to my comment, I have to yours 😅. If the comment says "removes tasks", but i just see it zeroing a count, I would be confused how that is "removing" a task. Which is why I am explaining what zeroing the task count really does.

File: src/code/z_skelanime.c (hunk 4 above)
Comment (fig02) [reply src/code/z_skelanime.c:815]: If I was learning about this system for the first time I personally would first look at functions that interface with the system, rather than the piece of data itself. Maybe its worth commenting on both idk

File: include/z64animation.h (hunk 1 above)
Comment (fig02) [reply include/z64animation.h:280]: That is exactly what it is yeah. I think a "copy map" is a good name for it, it maps out each limb to flag which denotes if it should be copied. I can make it longer with "limbCopyMap" if that helps. 
I already called the array in player a "limb copy map" in a previous round of documentation fwiw. but if we decide it should change then I can change it there too.

I personally dont find that "limb set" tells me much. set what? or do you mean a set of limbs?

File: include/z64animation.h (hunk 5 above)
Comment (fig02) [reply include/z64animation.h:289]: Yeah Im trying to convey its the inverse/reverse operation of "CopyUsingMap". I dont have any better ideas at the moment.
For what its worth this entry type is never used in practice by any actor, and isnt too important. That ofc does not mean we should neglect giving it a good name. Just putting it out there

Comment (cadmic) [reply include/z64animation.h:289]: It seems to be used by player? But yeah it's fine

File: include/z64animation.h (hunk 2 above)
Comment (fig02) [reply include/z64animation.h:261]: you can think of them as "regular" incrementing number sure, but the system never operates on it that way. The current group number is initialized with 1 (bit position 0) and incremented with `<<= 1`.
I would feel different if there was some interface with which a user can pass a group number as a "whole number" and the system does the shifting interpretation on its end. but this isnt the case.

It could be called a `groupBit` or something, but this is being more obtuse than needed imo, its still a single number that identifies a group, even if it doesnt count up normally.

File: src/code/z_skelanime.c (hunk 3 above)
Comment (fig02) [reply src/code/z_skelanime.c:803]: If this comment was reworded to this would it help:
"Because the queue only processes up to `count` number of tasks, zeroing `count` will stop the queue from processing old entries, effectively resetting the whole queue."

File: include/z64animation.h (hunk 1 above)
Comment (fig02) [reply include/z64animation.h:280]: yeah I think i still like calling it a copy map. does `limbCopyMap` help get my meaning across better?

File: src/code/z_skelanime.c (hunk 4 above)
Comment (fig02) [reply src/code/z_skelanime.c:815]: "Group numbers are indicated by a single bit that increments by shifting 1 position to the left. See `AnimTaskQueue_SetNextGroup` for more details"

would you be fine with this comment on the data member?

File: include/z64animation.h (hunk 5 above)
Comment (fig02) [reply include/z64animation.h:289]: Yeah that block of code is dead because `upperAnimBlendWeight` never gets set to non-zero

File: src/code/z_skelanime.c (hunk 3 above)
Comment (cadmic) [reply src/code/z_skelanime.c:803]: idk that still makes it seem like the rest of the array beyond `count` has valid tasks instead of being garbage, it implies that the old entries still exist and the word "effectively" makes it seem like an accident.

I'm kind of struggling to communicate my point here. Basically I think code comments should be high-level, saying what the code *means* instead of what the code *does* (since if you want to know what it does you can read the code). For example, if I was implementing a vector data structure I would not want comment it like this:

```c++
class IntVector {
 public:
  // Sets the count to 0.
  IntVector() {
    count = 0;
  }
   
  // Writes a new element to the array and adds 1 to the count.
  void push(int x) {
    data[count] = x;
    count++;
  }

  // Subtracts 1 from the count, effectively preventing any user from
  // processing the element that was at the back of the array before.
  void pop() {
    count--;
  }

  // Sets the count to 0, effectively preventing any user from processing
  // old entries until new ones are added.
  void clear() {
    count = 0;
  }
   
 private:
  int count;
  int data[MAX_SIZE];
}
```

Instead I would comment it like this:

```c++
class IntVector {
 public:
  // Creates an empty vector.
  IntVector() {
    count = 0;
  }
   
  // Appends an element to the end of the vector.
  void push(int x) {
    data[count] = x;
    count++;
  }

  // Removes an element from the end of the vector.
  void pop() {
    count--;
  }

  // Removes all elements from the vector.
  void clear() {
    count = 0;
  }
   
 private:
  int count;
  int data[MAX_SIZE];
}
```
Here the comments hit at what's going on conceptually, so the user can build a mental model of how to use the data structure, instead of saying how the code "happens" to work.

File: include/z64animation.h (hunk 1 above)
Comment (cadmic) [reply include/z64animation.h:280]: I think I do slightly prefer `limbCopyMap` since it hints that this is an array indexed by limb

File: include/z64animation.h (hunk 2 above)
Comment (cadmic) [reply include/z64animation.h:261]: Hmm I think `sDisableAnimQueueFlags`/`sDisabledTransformTaskGroups` is important here. It doesn't really make sense to or-together the different group ids; normally the only thing you're allowed to do with ids is compare them for equality. And the `if (!(task->group & sDisabledTransformTaskGroups))` is also pretty mysterious.

But it becomes clearer if you think of `sDisabledTransformTaskGroups` as a bitset where each bit represents a group i.e. group 0 is represented by bit 0, group 1 is represented by bit 1, and so on. Then `group` and `sCurAnimTaskGroup` are also group bitsets with one element set, or `(1 << group_num)` where `group_num` is the "real" group number (although it's never actually written down anywhere), and so `if (!(task->group & sDisabledTransformTaskGroups))` is testing whether the group number is contained in the set of disabled groups.

In summary: I still think this and `sCurAnimTaskGroup` should be thought of as `(1 << group_num)`, rather than single number that identifies a group, based on how `sDisabledTransformTaskGroups` is used. I think the names and comments should hint at that. Maybe `group` and `sCurAnimTaskGroup` is fine and my issue is more with the comment wording, the comments seem to treat this bit-twiddling as some kind of arcane implementation detail but it becomes easier to explain if you treat group numbers as bit numbers rather than powers of 2.

File: src/code/z_skelanime.c (hunk 4 above)
Comment (cadmic) [reply src/code/z_skelanime.c:815]: Maybe something general about group numbers would be helpful, but for the data members I'd prefer the comment to be more precise about what the data member itself represents, e.g.

```
// A bitset where the `i`th bit is set if the `i`th task group is disabled.
static u32 sDisabledTransformTaskGroups = 0;
// `(1 << i)` where `i` is the current task group.
static u32 sCurAnimTaskGroup;
```

Armed with this knowledge, the reader can guess that `sCurAnimTaskGroup <<= 1` is conceptually incrementing the "current task group", and something like `sCurAnimTaskGroup & sDisabledTransformTaskGroups` is testing whether the "current task group" is "disabled".

File: include/z64animation.h (hunk 2 above)
Comment (mzxrules) [reply include/z64animation.h:261]: ignoring the id aspect thing for a moment, I don't understand what's being "grouped"

Comment (fig02) [reply include/z64animation.h:261]: Different tasks or "entries" are being grouped together. Then, when `AnimTaskQueue_DisableTransformTasksForGroup` is called, the "transformative tasks" only for the currently active group (at the time of calling) will be skipped for that frame.

In practice, there are only two "groups" that get used. all of Links anim tasks are in one group, and all other actors tasks are in another. This lets Link disable the transformative tasks for his stuff, without affecting others. 
The same could be done in reverse, but `AnimTaskQueue_DisableTransformTasksForGroup` is only called by player.

File: src/code/z_skelanime.c (hunk 3 above)
Comment (fig02) [reply src/code/z_skelanime.c:803]: Im not implying that anything is an accident. "Effectively" here just means that doing this one action is what leads to something else occurring, this is its effect.
I give the high level explanation in the first sentence. I think its also fine to explain how the action taken leads to that desired high level effect. To me, the wording as-is is not a problem.

Someone else can weigh in if they'd like to see this changed as well

Comment (Dragorn421) [reply src/code/z_skelanime.c:803]: I agree with cadmic's points

I can suggest the following wording:
```
Clear the current task queue. The discarded tasks will then not be processed.
```

Comment (fig02) [reply src/code/z_skelanime.c:803]: When reading this code for the first time, I had to think for a moment how just zeroing the count leads to the queue being "reset" or tasks being discarded. Just saying the tasks are "discarded and wont be processed" doesn't help with my intent.
But point taken that the wording should change.

Comment (fig02) [reply src/code/z_skelanime.c:803]: maybe some version of my second sentence makes more sense as a comment in-line in the function? And the suggested comment can remain as the high-level function comment?

I do feel somewhat strongly about explaining this somewhere. It may seem obvious to you guys how it works, but it took me enough thought to feel like someone else could use the explanation as well.

Comment (Dragorn421) [reply src/code/z_skelanime.c:803]: Oh fwiw it seems straightforward to me
Or at least similar if not identical to a gazillion other queue/dynamic-array/buffer implementations

If you want to elaborate on how it works I'd put a comment inside the function (though I don't think it's necessary)

Comment (fig02) [reply src/code/z_skelanime.c:803]: alright i will concede and just not explain it

File: tools/disasm/gc-eu-mq/functions.txt
Comment (cadmic) [inline tools/disasm/gc-eu-mq/functions.txt:1]: this file should not be added (bad merge?)

**Review by cadmic — APPROVED** [review]
I don't feel strongly about the "group" discussion above and I think this is still an improvement

**Comment by fig02** [comment]
fwiw I was going to make changes and try to find a compromise that works for the both of us. just havent found the time or motivation.

I'd be okay with merging this as-is, and then further changes can be made if people find it necessary. I mainly want to move on to documenting the "Anim Move" stuff on Player's side. That was the whole point of doing this anyway.

**Comment by cadmic** [comment]
Yeah I personally think player docs are more important

**Review by mzxrules — APPROVED** [review]
I'm OK with merging this for now (once the red x is fixed of course).
